helm/qontract-reconcile/templates/template.yaml

```diff
@@ -127,8 +127,102 @@ objects:
                 key: log_group_name
           command: ["/bin/sh", "-c"]
           args:
+          # generate /etc/vector/vector.yaml
+          # and generate /fluentd/etc/fluent.conf
           - |
-            # generate fluent.conf
+            ### APPSRE-8096: Vector Section Begins
+            mkdir -p /etc/vector
+            cat > /etc/vector/vector.yaml <<EOF
+            sources:
+              qr:
+                type: file
+                include:
+                  - /fluentd/log/integration.log
+                read_from: beginning
+                data_dir: /fluentd/log
+
+            transforms:
+              ### De-spaminate
+              despam_events:
+                type: filter
+                inputs:
+                  - qr
+                condition:
+                  type: datadog_search
+                  source: -"using gql endpoint" AND -"Certificate did not match expected hostname"
+
+              ### Craft a JSON document that has escaped double quotes and changes the event from the 
+              ### Vector-default `message` to `text`, which is require by Slack and Google Chat.
+              rename_message_webhooks:
+                type: remap
+                inputs:
+                  - despam_events
+                source: |-
+                  .message = join!(
+                  [
+                    "{\"text\":\"",
+                    format_timestamp!(
+                      now(),
+                      "%Y-%m-%d %H:%M:%S %z"
+                    ),
+                    " ",
+                    " [${POD_NAME}] ",
+                    replace!(
+                      del(.message),
```

Comment:
Okay this one is kinda throwing me through a loop as I'm understanding VRL. So this basically means that the `.message` field is removed from this event and returned from the `del` function. And the end result is that `.message` will equal the result of this join? Right?

Reply:
Not quite. Let me annotate the entirety of the `.message = join!(....)`:

```
161 .message = join!(
162 [
163   "{\"text\":\"",
164   format_timestamp!(
165     now(),
166     "%Y-%m-%d %H:%M:%S %z"
167   ),
168   " ",
169   " [${POD_NAME}] ",
170   replace!(
171     del(.message),
172     "\"",
173     "\\\""),
174   "\"}"
175 ])
```

Example log message from /var/log/system.log
```
Jul 25 11:36:39 rhodium syslogd[143]: ASL Sender Statistics
```

Line 161: By default, when Vector emits a logging event, the pathway from the "source" to the "sink(s)" involves creating a sort of object on which is a `message` attribute. This `message` attribute is the content of the event, such as a log line. In the example, `message` would be equal to `ASL Sender Statistics`.

This poses a problem because when using the webhooks of Slack and Google Chat to transmit the log message, each endpoint wants the payload to be a JSON document in the format of `{"text": "ASL Sender Statistics"}`. Additionally, the sink wants the source of that payload to be the contents of the `message` attribute. That means that we must "transform" the `message` to some other value.

The way to construct the JSON payload is to join a bunch of strings together into one final payload document, and that begins on this line 161 by reassigning the value of the `message` attribute to the result of a `join`. The `!` is required to discard any potential errors that _could_ occur (as far as Vector is concerned) in the `join` process.

Line 162: Declare the start of the array that will be joined.

Line 163: Escape double quotes to set the first element of the array to `{"text":"}`. Note the trailing `"` which declares the start of the JSON value string for `text` JSON key.

Line 164-167: Format the current timestamp with the standard `strftime(3)` expansions. Here, again, the `!` tells Vector to dicard any possible errors from the `format_timestamp` method. The reulting string is the second element in the array-to-be-joined, and represents the first thing in the `text` payload string.

Line 168: Insert a space immediately after the timestamp as the third array element.

Line 169: Reference and interpolate the environment variable, `POD_NAME`, (which is set in the `PodSpec`) and include that in the `text` payload.

Line 170-173: If the `message` contained a doublequote (`"`) character, it would break the construction of the JSON document we're making. That means we must globally `replace` every `"` in the `message` with an escaped version of it.

Line 170: The `replace!` method takes the format of `replace!(haystack, needle, replacement)` where all needles in the haystack are replaced. As before, the `!` discards any potential errors from the `replace` method.

Line 171: `del(.message)` deletes _and returns_ from the object the `message`. This return value serves as an input into the `replace!` method on line 170.

Line 172: Search for literal `"`

Line 173: Replace them with a literal `\"`

Line 174: The final element in the join element is `"}`, thus closing the opened string on line 163.

Line 175: Terminate the array, and close the arguments into the `join!` call from line 161.

Reply:
Thank you for taking the time to explain all of this. Makes sense to me now

Reply:
👍 I'll resolve the convo, but link this comment thread in the main PR text.